lib/collection/src/shards/local_shard/scroll.rs

```diff
@@ -219,4 +242,47 @@ impl LocalShard {
 
         Ok((ordered_records, values))
     }
+
+    async fn scroll_randomly(
+        &self,
+        limit: usize,
+        with_payload_interface: &WithPayloadInterface,
+        with_vector: &WithVector,
+        filter: Option<&Filter>,
+        search_runtime_handle: &Handle,
+    ) -> CollectionResult<Vec<Record>> {
+        let segments = self.segments();
+
+        let (non_appendable, appendable) = segments.read().split_segments();
+
+        let read_filtered = |segment: LockedSegment| {
+            let filter = filter.cloned();
+
+            search_runtime_handle.spawn_blocking(move || {
+                segment
+                    .get()
+                    .read()
+                    .read_random_filtered(limit, filter.as_ref())
+            })
+        };
+
+        let all_reads = try_join_all(
+            non_appendable
+                .into_iter()
+                .map(read_filtered)
+                .chain(appendable.into_iter().map(read_filtered)),
+        )
+        .await?;
+
+        let point_ids = all_reads
+            .into_iter()
+            .flatten()
+            .choose_multiple(&mut rand::thread_rng(), limit);
```

Comment:
IIRC, though this works, it can be very expensive as it might evaluate all possible values in the iterator. I must admit that I may be wrong here, as the details on it's implementation are quite confusing.

I wonder if the following would be better: create a list of all random readers, so you have one for each segment. Fill our final list by constantly picking from a random reader. This will break uniformity though with segments that don't have equal size.

Also, we should deduplicate here. It is possible for different segments to have the same point ID. Ideally we'd only take the one with the highest version, but I don't see how we can do that here.

Do you have any thoughts on this based on the above?

Reply:
Thank you, I did forget to deduplicate 👍.

Implementation of `choose_multiple` is indeed `O(n)`. I did try to do the streaming solution, but was kind of defeated for the day. I'll give it another go, it should be better than fetching the whole limit from every shard, since we don't need any kind of "best result" here.